src/control/jobs/control_jobs.c

```diff
@@ -1829,50 +1830,58 @@ void dt_control_export(GList *imgid_list, int max_width, int max_height, int for
   mstorage->export_dispatched(mstorage);
 }
 
-static int32_t dt_control_time_offset_job_run(dt_job_t *job)
+static int32_t dt_control_datetime_job_run(dt_job_t *job)
 {
   dt_control_image_enumerator_t *params = (dt_control_image_enumerator_t *)dt_control_job_get_params(job);
   uint32_t cntr = 0;
   double fraction = 0.0;
   GList *t = params->index;
-  const long int offset = ((dt_control_time_offset_t *)params->data)->offset;
+  const long int offset = ((dt_control_datetime_t *)params->data)->offset;
+  const char *datetime = ((dt_control_datetime_t *)params->data)->datetime;
   char message[512] = { 0 };
 
   /* do we have any selected images and is offset != 0 */
-  if(!t || offset == 0)
+  if(!t || (offset == 0 && !datetime[0]))
   {
     return 1;
   }
 
   const guint total = g_list_length(t);
 
-  snprintf(message, sizeof(message),
-           ngettext("adding time offset to %d image", "adding time offset to %d images", total), total);
+  const char *mes1 = offset ? "adding time offset to %d image" : "setting date time to %d image";
+  snprintf(message, sizeof(message), ngettext(mes1, mes1, total), total);
   dt_control_job_set_progress_message(job, message);
 
+  dt_undo_start_group(darktable.undo, DT_UNDO_DATETIME);
   /* go thru each selected image and update datetime_taken */
   do
   {
     int imgid = GPOINTER_TO_INT(t->data);
 
-    dt_image_add_time_offset(imgid, offset);
+    if(offset)
+      dt_image_add_time_offset(imgid, offset);
+    else
+      dt_image_set_datetime(imgid, datetime);
     cntr++;
 
     fraction = MAX(fraction, (1.0 * cntr) / total);
     dt_control_job_set_progress(job, fraction);
   } while((t = g_list_next(t)) != NULL);
+  dt_undo_end_group(darktable.undo);
 
-  dt_control_log(ngettext("added time offset to %d image", "added time offset to %d images", cntr), cntr);
-
+  const char *mes2 = offset ? "added time offset to %d image" : "set date time to %d image";
```

Comment:
You've removed the plural forms. And the strings need to be placed into `N_(" ... ")` for them to enter the translation catalog and for the below ngettext() call to be able to find them.

Reply:
Sorry I have not seen them ...

Reply:
Done.